bokehjs/src/lib/models/annotations/dimensional.ts

```diff
@@ -0,0 +1,210 @@
+    assert(found_unit != null)
+
+    const value_in_unit = value*found_unit.factor
+
+    const [new_unit, new_long_unit, new_value] = (() => {
+      const index = bisect_right_by(basis, value_in_unit, ({factor}) => factor)
+      if (index > 0) {
+        const {short_name: new_unit, long_name: new_long_unit, factor} = basis[index - 1]
+        const new_value = value_in_unit/factor
+        return [new_unit, new_long_unit, new_value]
+      } else {
+        return [unit, found_unit.long_name, value_in_unit]
+      }
+    })()
+
+    exact = exact ?? ticks.length == 0
+
+    const preferred_value = (() => {
+      if (exact) {
+        return new_value
+      } else {
+        const index = bisect_right(ticks, new_value)
+        return ticks[min(index, ticks.length-1)]
+      }
+    })()
+
+    const preferred_value_raw = preferred_value*(value_in_unit/new_value)
+    const scale_factor = (preferred_value_raw/value)/found_unit.factor
+
+    return {
+      new_value: preferred_value,
+      new_unit,
+      new_long_unit,
+      scale_factor,
+      exact,
+    }
+  }
+}
+
+export namespace Metric {
+  export type Attrs = p.AttrsOf<Props>
+  export type Props = Dimensional.Props
+}
+
+export interface Metric extends Metric.Attrs {}
+
+export abstract class Metric extends Dimensional {
+  declare properties: Metric.Props
+
+  constructor(attrs?: Partial<Metric.Attrs>) {
+    super(attrs)
+  }
+
+  static {
+    this.override<Metric.Props>({
+      ticks: [1, 2, 5, 10, 15, 20, 25, 50, 75, 100, 125, 150, 200, 250, 500, 750],
+    })
+  }
+
+  build_basis(): Basis {
+    const {_basis_template, _short_name, _long_name, _tex_repr} = this
+    const {include, exclude} = this
+
+    const basis = _basis_template.map(([short_prefix, factor, long_prefix, tex_prefix]) => {
+      return {
+        short_name: `${short_prefix}${_short_name}`,
+        factor,
+        long_name: `${long_prefix}${_long_name}`,
+        tex_repr:  `${tex_prefix ?? short_prefix}${_tex_repr}`,
+      }
+    }).filter(({short_name}) => {
+      return (include == null || include.includes(short_name)) && !exclude.includes(short_name)
+    })
+
+    return sort_by(basis, ({factor}) => factor)
+  }
+
+  protected _basis: Basis | null = null
+  get basis(): Basis {
+    if (this._basis == null) {
+      this._basis = this.build_basis()
+    }
+    return this._basis
+  }
+
+  protected abstract _short_name: string
+  protected abstract _long_name: string
+  protected abstract _tex_repr: string
+
+  protected _basis_template: [short_name: string, factor: number, long_name: string, tex: string | null][] = [
+    ["Q", 1e30,  "quetta", null],
+    ["R", 1e27,  "ronna", null],
+    ["Y", 1e24,  "yotta", null],
+    ["Z", 1e21,  "zetta", null],
+    ["E", 1e18,  "exa", null],
+    ["P", 1e15,  "peta", null],
+    ["T", 1e12,  "tera", null],
+    ["G", 1e9,   "giga", null],
+    ["M", 1e6,   "mega", null],
+    ["k", 1e3,   "kilo", null],
+    ["h", 1e2,   "hecto", null],
```

Comment:
I would probably opt for just powers of 3 here, so avoiding `hecto`, `deci` and `centi`, but I don't have a strong opinion.

Ah, I see that two of these three are by default excluded below.

Reply:
Yes, this is configurable and the first two are excluded by default from `MetricLength`.